src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,280 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
+        all_waveforms = np.lib.format.open_memmap(file_path, mode="w+", dtype=dtype, shape=shape)
+        wf_array_info = str(file_path)
+    elif mode == "shared_memory":
+        if num_spikes == 0 or num_chans == 0:
+            all_waveforms = np.zeros(shape, dtype=dtype)
+            shm = None
+            shm_name = None
+        else:
+            all_waveforms, shm = make_shared_array(shape, dtype)
+            shm_name = shm.name
+        wf_array_info = (shm, shm_name, dtype.str, shape)
+    else:
+        raise ValueError("allocate_waveforms_buffers bad mode")
+
+    job_kwargs = fix_job_kwargs(job_kwargs)
+
+    if num_spikes > 0:
+        # and run
+        func = _worker_distribute_single_buffer
+        init_func = _init_worker_distribute_single_buffer
+
+        init_args = (
+            recording,
+            unit_ids,
+            spikes,
+            wf_array_info,
+            nbefore,
+            nafter,
+            return_scaled,
+            mode,
+            sparsity_mask,
+        )
+        if job_name is None:
+            job_name = f"extract waveforms {mode} mono buffer"
+
+        processor = ChunkRecordingExecutor(recording, func, init_func, init_args, job_name=job_name, **job_kwargs)
+        processor.run()
+
+    if mode == "memmap":
+        return all_waveforms
+    elif mode == "shared_memory":
+        if copy:
+            if shm is not None:
+                # release all sharedmem buffer
+                # empty array have None
+                shm.unlink()
+            return all_waveforms.copy()
+        else:
+            return all_waveforms, wf_array_info
+
+
+def _init_worker_distribute_single_buffer(
+    recording, unit_ids, spikes, wf_array_info, nbefore, nafter, return_scaled, mode, sparsity_mask
+):
+    worker_ctx = {}
+    worker_ctx["recording"] = recording
+    worker_ctx["wf_array_info"] = wf_array_info
+    worker_ctx["spikes"] = spikes
+    worker_ctx["nbefore"] = nbefore
+    worker_ctx["nafter"] = nafter
+    worker_ctx["return_scaled"] = return_scaled
+    worker_ctx["sparsity_mask"] = sparsity_mask
+    worker_ctx["mode"] = mode
+
+    if mode == "memmap":
+        filename = wf_array_info
+        all_waveforms = np.load(str(filename), mmap_mode="r+")
+        worker_ctx["all_waveforms"] = all_waveforms
+    elif mode == "shared_memory":
+        from multiprocessing.shared_memory import SharedMemory
+
+        shm, shm_name, dtype, shape = wf_array_info
+        shm = SharedMemory(shm_name)
+        all_waveforms = np.ndarray(shape=shape, dtype=dtype, buffer=shm.buf)
+        worker_ctx["shm"] = shm
+        worker_ctx["all_waveforms"] = all_waveforms
+
+    # prepare segment slices
+    segment_slices = []
+    for segment_index in range(recording.get_num_segments()):
+        s0 = np.searchsorted(spikes["segment_index"], segment_index)
+        s1 = np.searchsorted(spikes["segment_index"], segment_index + 1)
+        segment_slices.append((s0, s1))
```

Comment:
I think this is the spot where it will easier to filter for spikes whose waveform does not lie entirely in the traces with the choosen nbefore and nafter.

```python
spikes_in_segment = spikes[s0:s1]
spikes_left_bounds = spikes_in_segment["sample_index"] - nbefore,
num_spikes_with_invalid_left_bound = np.searchsorted(spikes_left_bounds, 0)
s0 += num_spikes_with_invalid_left_bound
spikes_right_bounds = spikes_in_segment["sample_index"] + nafter,
num_spikes_with_invalid_right_bound = np.searchsorted(spikes_right_bounds, num_frames)
s1 -= num_spikes_with_invalid_right_bound
s1 = np.searchsorted(spikes["segment_index"], segment_index + 1)

```

Reply:
I did another trick finally which is similar and easy.

Reply:
Great!